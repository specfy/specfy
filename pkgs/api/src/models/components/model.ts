import type { Prisma, Activities, Components, Users } from '@prisma/client';

import { nanoid } from '../../common/id.js';
import { slugify } from '../../common/string.js';
import type { ActionComponent } from '../activities/types.js';

export async function createComponentBlob({
  data,
  blob,
  tx,
}: {
  data?: Partial<Pick<Prisma.BlobsCreateInput, 'created' | 'deleted'>>;
  blob: Components | Prisma.ComponentsUncheckedCreateInput;
  tx: Prisma.TransactionClient;
}) {
  return await tx.blobs.create({
    data: {
      id: nanoid(),
      parentId: blob.blobId || null,
      type: 'component',
      typeId: blob.id,
      current: blob as any,
      created: false,
      deleted: false,
      ...data,
    },
  });
}

export async function createComponent({
  data,
  user,
  tx,
}: {
  data: Omit<Prisma.ComponentsUncheckedCreateInput, 'blobId' | 'slug'>;
  user: Users;
  tx: Prisma.TransactionClient;
}) {
  const body: Prisma.ComponentsUncheckedCreateInput = {
    show: true,
    tags: [],
    ...data,
    slug: slugify(data.name),
    id: data.id || nanoid(),
    blobId: null,
  };
  const blob = await createComponentBlob({
    data: { created: true },
    blob: body,
    tx,
  });
  const model: Prisma.ComponentsUncheckedCreateInput = {
    ...body,
    blobId: blob.id,
  };

  const tmp = await tx.components.create({
    data: model,
  });
  await createComponentActivity({
    user,
    action: 'Component.created',
    target: tmp,
    tx,
  });

  return tmp;
}

export async function createComponentActivity({
  user,
  action,
  target,
  tx,
  activityGroupId = null,
}: {
  user: Users;
  action: ActionComponent;
  target: Components;
  tx: Prisma.TransactionClient;
  activityGroupId?: string | null;
}): Promise<Activities> {
  return await tx.activities.create({
    data: {
      id: nanoid(),
      action,
      userId: user.id,
      orgId: target.orgId,
      projectId: target.projectId,
      activityGroupId,
      targetBlobId: target.blobId,
      createdAt: new Date(),
    },
  });
}
