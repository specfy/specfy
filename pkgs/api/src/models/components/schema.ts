import { z } from 'zod';

import {
  schemaId,
  schemaSlug,
  schemaOrgId,
} from '../../common/validators/index.js';
import { schemaProseMirror } from '../../common/validators/prosemirror.js';
import { schemaDisplay, schemaEdges } from '../flows/validator.js';

export const schemaComponent = z
  .object({
    id: schemaId,
    orgId: schemaOrgId,
    projectId: schemaId,
    blobId: schemaId.nullable(),
    techId: z.string().nonempty().nullable(), // TODO: do something about that?

    // Can't find a way to sync them
    // type: z.enum([...componentTypes]),
    type: z.enum([
      'app',
      'ci',
      'db',
      'hosting',
      'language',
      'messaging',
      'network',
      'saas',
      'storage',
      'tool',
      'project',
      'service',
    ]),
    typeId: schemaId.nullable(),

    name: z.string().min(2).max(100),
    slug: schemaSlug,
    description: schemaProseMirror,
    techs: z.string().max(25).array(), // TODO: do something about that?

    display: schemaDisplay,
    edges: schemaEdges,

    inComponent: schemaId.nullable(),

    source: z.string().max(100).nullable(),
    sourceName: z.string().max(100).nullable(),
    sourcePath: z.array(z.string().max(500)).max(50).nullable(),

    tags: z.string().max(20).array(),
    show: z.boolean(),

    createdAt: z.string().datetime(),
    updatedAt: z.string().datetime(),
  })
  .strict();
