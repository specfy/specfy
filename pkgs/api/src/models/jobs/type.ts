import type { Prisma } from '@prisma/client';

export interface JobDeployConfig {
  url: string;
  autoLayout?: boolean;
  hook?: {
    id: string;
    ref: string;
  };
}

export type JobWithOrgProject = Prisma.JobsGetPayload<{
  include: { Org: true; Project: true };
}>;

export type JobCode =
  | 'cant_auth_github'
  | 'failed_to_checkout'
  | 'failed_to_cleanup'
  | 'failed_to_clone'
  | 'failed_to_deploy'
  | 'failed_to_start_github_deployment'
  | 'failed_to_teardown'
  | 'missing_dependencies'
  | 'no_api_key'
  | 'org_not_installed'
  | 'project_not_installed'
  | 'success'
  | 'unknown';
