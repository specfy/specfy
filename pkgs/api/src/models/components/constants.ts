import type { ComponentType } from './types.js';

export const componentTypes: ComponentType[] = [
  'app',
  'ci',
  'db',
  'hosting',
  'language',
  'messaging',
  'network',
  'sass',
  'storage',
  'tool',
  'project',
  'service',
];
