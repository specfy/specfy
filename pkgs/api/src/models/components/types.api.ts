import type {
  Pagination,
  QuerystringOrgProject,
  Res,
} from '../../types/api/api.js';

import type { DBComponent } from './types.js';

export type ApiComponent = DBComponent;

// GET /
export type ListComponents = Res<{
  Querystring: QuerystringOrgProject;
  Success: {
    data: ApiComponent[];
    pagination: Pagination;
  };
}>;
