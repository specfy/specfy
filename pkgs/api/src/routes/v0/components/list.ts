import type { FastifyPluginCallback, FastifyRequest } from 'fastify';
import { z } from 'zod';

import { validationError } from '../../../common/errors.js';
import { schemaId, schemaOrgId } from '../../../common/validators/index.js';
import { valPermissions } from '../../../common/zod.js';
import { prisma } from '../../../db/index.js';
import type { ListComponents, Pagination } from '../../../types/api/index.js';

function QueryVal(req: FastifyRequest) {
  return z
    .object({
      org_id: schemaOrgId,
      project_id: schemaId,
    })
    .strict()
    .superRefine(valPermissions(req));
}

const fn: FastifyPluginCallback = (fastify, _, done) => {
  fastify.get<ListComponents>('/', async function (req, res) {
    const val = QueryVal(req).safeParse(req.query);
    if (!val.success) {
      return validationError(res, val.error);
    }

    const query: ListComponents['Querystring'] = val.data;

    // TODO: pagination or remove it
    const pagination: Pagination = {
      currentPage: 0,
      totalItems: 0,
    };

    const docs = await prisma.components.findMany({
      where: {
        orgId: query.org_id,
        projectId: query.project_id,
      },
      orderBy: { name: 'asc' },
      take: 1000,
      skip: 0,
    });

    return res.status(200).send({
      data: docs.map((p) => {
        // For excess property check
        const tmp: ListComponents['Success']['data'][0] = {
          id: p.id,
          orgId: p.orgId,
          projectId: p.projectId,
          blobId: p.blobId,
          techId: p.techId,

          type: p.type,
          typeId: p.typeId,
          name: p.name,
          slug: p.slug,
          description: p.description,
          techs: p.techs,
          display: p.display,
          inComponent: p.inComponent,
          edges: p.edges,

          show: p.show,
          tags: p.tags,

          source: p.source,
          sourceName: p.sourceName,
          sourcePath: p.sourcePath,

          createdAt: p.createdAt.toISOString(),
          updatedAt: p.updatedAt.toISOString(),
        };
        return tmp;
      }),
      pagination,
    });
  });
  done();
};

export default fn;
