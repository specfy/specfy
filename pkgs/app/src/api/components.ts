import type { ApiComponent, ListComponents } from '@specfy/api/src/types/api';
import { useQuery } from '@tanstack/react-query';

import originalStore from '../common/store';

import { fetchApi } from './fetch';
import { APIError, isError } from './helpers';

export function useListComponents(
  slug: string,
  opts: Partial<ListComponents['Querystring']>
) {
  return useQuery({
    enabled: !!(opts.org_id && opts.project_id),
    queryKey: ['listComponents', opts.org_id, slug],
    queryFn: async (): Promise<ApiComponent[]> => {
      const { json, res } = await fetchApi<ListComponents>('/components', {
        qp: { org_id: opts.org_id!, project_id: opts.project_id! },
      });

      if (res.status !== 200 || isError(json)) {
        throw new APIError({ res, json });
      }

      return json.data.map((component) => {
        originalStore.add(component);
        return component;
      });
    },
  });
}
