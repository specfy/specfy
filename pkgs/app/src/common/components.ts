import { nanoid } from '@specfy/api/src/common/id';
import {
  hDef,
  hDefHost,
  wDef,
  wDefHost,
} from '@specfy/api/src/common/validators/flow.constants';
import type { ApiComponent } from '@specfy/api/src/types/api';
import type { FlowEdge } from '@specfy/api/src/types/db';

import { getEmptyDoc } from './content';
import type { ComponentsState, ProjectState } from './store';

/**
 * Get all childs of a component.
 */
export function getAllChilds(
  components: ApiComponent[],
  id: string
): ApiComponent[] {
  const tmp = [];
  for (const component of components) {
    if (component.inComponent === id) {
      tmp.push(component);
      tmp.push(...getAllChilds(components, component.id));
    }
  }
  return tmp;
}

export function createLocal(
  data: Partial<Pick<ApiComponent, 'techId' | 'typeId'>> &
    Pick<ApiComponent, 'name' | 'slug' | 'type'> & {
      position?: { x: number; y: number };
    },
  storeProject: ProjectState,
  storeComponents: ComponentsState
) {
  const id = nanoid();

  const size =
    data.type === 'hosting'
      ? { width: wDefHost, height: hDefHost }
      : { width: wDef, height: hDef };

  let pos = { x: 0, y: 0 };

  // Compute global bounding box
  if (data.position) {
    pos = data.position;
  } else {
    const global = { x: 0, y: 0, width: 0, height: 0 };
    for (const component of Object.values(storeComponents.components)) {
      global.x = Math.min(component.display.pos.x, global.x);
      global.y = Math.min(component.display.pos.y, global.y);
      global.width = Math.max(component.display.size.width, global.width);
      global.height = Math.max(component.display.size.height, global.height);
    }

    // Simply add on top of it
    pos = { x: global.x, y: global.y - size.height };
  }

  const add: ApiComponent = {
    id,
    orgId: storeProject.project!.id,
    projectId: storeProject.project!.id,
    techId: data.techId || null,
    type: data.type,
    typeId: data.typeId || null,
    name: data.name,
    slug: data.slug,
    description: getEmptyDoc(),
    techs: [],
    display: { pos: pos, size },
    edges: [],
    blobId: '',
    inComponent: null,
    show: true,
    tags: [],
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
  };

  storeComponents.create(add);
  return id;
}

/**
 * Position the edge on port from a to b.
 */
export function positionEdge(
  a: ApiComponent,
  b: ApiComponent
): { source: FlowEdge['portSource']; target: FlowEdge['portTarget'] } {
  const isCurrentAbove =
    a.display.pos.y + a.display.size.height < b.display.pos.y;
  const isCurrentBelow =
    a.display.pos.y > b.display.pos.y + b.display.size.height;
  const isCurrentRight =
    a.display.pos.x > b.display.pos.x + b.display.size.width;
  const isCurrentLeft = a.display.pos.x + a.display.pos.x < b.display.pos.x;

  let source: FlowEdge['portSource'] = 'left';
  let target: FlowEdge['portTarget'] = 'right';
  if (isCurrentLeft) {
    source = 'right';
    target = 'left';
  } else if (isCurrentAbove && !isCurrentRight) {
    source = 'bottom';
    target = 'top';
  } else if (isCurrentBelow) {
    source = 'top';
    target = 'bottom';
  }

  return { source, target };
}
