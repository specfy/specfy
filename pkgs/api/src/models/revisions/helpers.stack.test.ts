import type { Components, Orgs, Projects } from '@prisma/client';
import type { AnalyserJson } from '@specfy/stack-analyser';
import { Payload } from '@specfy/stack-analyser';
import { describe, expect, it } from 'vitest';

import { getBlobComponent } from '../../test/seed/components.js';
import type { PostUploadRevision } from '../../types/api/index.js';

import { uploadedStackToDB } from './helpers.stack.js';

function getDefault(): AnalyserJson {
  return new Payload({ name: 'default', folderPath: '/' }).toJson();
}
describe('uploadedStackToDB', () => {
  it('should output new blobs once and nothing the second', () => {
    // New upload
    const res = uploadedStackToDB(
      {
        ...getDefault(),
        childs: [
          { ...getDefault(), name: 'a' },
          { ...getDefault(), name: 'b' },
        ],
      },
      [],
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res.blobs).toHaveLength(2);
    expect(res.deleted).toHaveLength(0);
    const snap = res.blobs.map((b) => {
      const copy = {
        ...b,
        id: expect.any(String),
        typeId: expect.any(String),
        current: {
          ...b.current,
          id: expect.any(String),
          createdAt: '',
          updatedAt: '',
        },
      };
      return copy;
    });
    expect(snap).toMatchSnapshot();

    // Re-upload
    const res2 = uploadedStackToDB(
      {
        ...getDefault(),
        childs: [
          { ...getDefault(), name: 'a' },
          { ...getDefault(), name: 'b' },
        ],
      },
      res.blobs.map((b) => b.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(2);
    expect(res2.deleted).toHaveLength(0);
    expect(res2.blobs[0].typeId).toEqual(res.blobs[0].typeId);
    expect(res2.blobs[1].typeId).toEqual(res.blobs[1].typeId);
    expect(res2.unchanged).toStrictEqual([
      res2.blobs[0].typeId,
      res2.blobs[1].typeId,
    ]);
  });

  it('should update id/edges.id when one component was already there', () => {
    // New upload
    const up: AnalyserJson = {
      ...getDefault(),
      name: 'coucou',
      tech: 'algolia',
    };
    const first = uploadedStackToDB(
      {
        ...getDefault(),
        childs: [up],
      },
      [],
      { source: 'github' } as PostUploadRevision['Body']
    );

    // Re-upload
    const reup: AnalyserJson = {
      ...getDefault(),
      name: 'coucou',
      tech: 'algolia',
    };
    const res2 = uploadedStackToDB(
      {
        ...getDefault(),
        childs: [
          reup,
          {
            ...getDefault(),
            name: 'b',
            edges: [
              {
                target: reup.id,
                read: true,
                write: true,
              },
            ],
          },
        ],
      },
      first.blobs.map((b) => b.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(2);
    expect(res2.blobs[0].typeId).toEqual(up.id);
    expect(res2.blobs[0].current.id).toEqual(up.id);
    expect(res2.blobs[1].current.edges[0].target).toEqual(up.id);
    expect(res2.unchanged).toStrictEqual([res2.blobs[0].typeId]);
  });

  it('should detect deleted', () => {
    // New upload
    const up: AnalyserJson = {
      ...getDefault(),
      name: 'coucou',
      tech: 'algolia',
    };
    const first = uploadedStackToDB(
      {
        ...getDefault(),
        childs: [up],
      },
      [],
      { source: 'github' } as PostUploadRevision['Body']
    );

    // Re-upload
    const res2 = uploadedStackToDB(
      {
        ...getDefault(),
        childs: [],
      },
      first.blobs.map((b) => b.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(0);
    expect(res2.deleted).toHaveLength(1);
    expect(res2.deleted[0].typeId).toEqual(up.id);
    expect(res2.unchanged).toStrictEqual([]);
  });

  it('should not delete different source', () => {
    // New upload
    const up: AnalyserJson = {
      ...getDefault(),
      name: 'coucou',
      tech: 'algolia',
    };
    const res = uploadedStackToDB(
      {
        ...getDefault(),
        childs: [up],
      },
      [
        {
          ...(getBlobComponent(
            { id: 'acme' } as Orgs,
            { id: 'foobar' } as Projects
          ) as unknown as Components),
        },
      ],
      { source: 'github' } as PostUploadRevision['Body']
    );

    expect(res.deleted).toHaveLength(0);
  });
});
