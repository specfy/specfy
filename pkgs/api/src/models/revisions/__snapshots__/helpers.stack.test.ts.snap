// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uploadedStackToDB > should output new blobs once and nothing the second 1`] = `
[
  {
    "created": true,
    "current": {
      "blobId": null,
      "createdAt": "",
      "description": {
        "content": [],
        "type": "doc",
      },
      "display": {
        "pos": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 32,
          "width": 100,
        },
        "zIndex": 1,
      },
      "edges": [],
      "id": Any<String>,
      "inComponent": null,
      "name": "A",
      "orgId": undefined,
      "projectId": undefined,
      "show": true,
      "slug": "a",
      "source": "github",
      "sourceName": "a",
      "sourcePath": [
        "/",
      ],
      "tags": [
        "github",
      ],
      "techId": null,
      "techs": [],
      "type": "service",
      "typeId": null,
      "updatedAt": "",
    },
    "deleted": false,
    "id": Any<String>,
    "parentId": null,
    "type": "component",
    "typeId": Any<String>,
  },
  {
    "created": true,
    "current": {
      "blobId": null,
      "createdAt": "",
      "description": {
        "content": [],
        "type": "doc",
      },
      "display": {
        "pos": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 32,
          "width": 100,
        },
        "zIndex": 1,
      },
      "edges": [],
      "id": Any<String>,
      "inComponent": null,
      "name": "B",
      "orgId": undefined,
      "projectId": undefined,
      "show": true,
      "slug": "b",
      "source": "github",
      "sourceName": "b",
      "sourcePath": [
        "/",
      ],
      "tags": [
        "github",
      ],
      "techId": null,
      "techs": [],
      "type": "service",
      "typeId": null,
      "updatedAt": "",
    },
    "deleted": false,
    "id": Any<String>,
    "parentId": null,
    "type": "component",
    "typeId": Any<String>,
  },
]
`;
